### PR TITLE
[AssignmentTracking] Skip large types in redundant debug info pruning

### LINKED ISSUE #74189: Clang-18 crash when compiled with -O2 -g -c.
Clang-18 with -O2 -g -c crashes on the test case.
Compiler explorer: https://godbolt.org/z/MsheKq9vb
```
$cat mutant.c
struct __attribute__((packed)) A { char c1; short a1[__INT_MAX__]; };
struct __attribute__((packed)) B { char c2; short a2[100]; };
struct S { struct A p1; struct B p2[4]; };
void bar (short int);

static void
foo (struct S *q)
{
  for (int i = 0; i < q->p1.c1; i++)
    for (int j = 0; j < q->p2[i].c2; j++)
      bar (q->p2[i].a2[j]);
}

int
main ()
{
  struct S q = {};
  q.p2[0].c2 = q.p2[1].c2 = 3;
  foo (&q);
}

$clang-18 -O2 -g -c mutant.c
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang-18 -O2 -g -c mutant.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'mutant.c'.
4.      Running pass 'Assignment Tracking Analysis' on function '@main'
 #0 0x00007ff48329b266 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd2b266)
 #1 0x00007ff483299170 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd29170)
 #2 0x00007ff48329a8c4 llvm::sys::CleanupOnSignal(unsigned long) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xd2a8c4)
 #3 0x00007ff4831e8bb0 (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xc78bb0)
 #4 0x00007ff48dd7c980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #5 0x00007ff4834b6e2a (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xf46e2a)
 #6 0x00007ff4834b2418 llvm::AssignmentTrackingAnalysis::runOnFunction(llvm::Function&) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xf42418)
 #7 0x00007ff4833e10df llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xe710df)
 #8 0x00007ff4833e6dc3 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xe76dc3)
 #9 0x00007ff4833e17d4 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xe717d4)
#10 0x00007ff48ba7e09f clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x1ade09f)
#11 0x00007ff48be145fe (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x1e745fe)
#12 0x00007ff48aa30866 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0xa90866)
#13 0x00007ff48c888645 clang::FrontendAction::Execute() (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x28e8645)
#14 0x00007ff48c806cc4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x2866cc4)
#15 0x00007ff48c9031c0 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x29631c0)
#16 0x0000561c05ddd837 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/lib/llvm-18/bin/clang+0x13837)
#17 0x0000561c05dda905 (/usr/lib/llvm-18/bin/clang+0x10905)
#18 0x00007ff48c49e909 (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x24fe909)
#19 0x00007ff4831e894c llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/llvm-18/bin/../lib/libLLVM-18.so.1+0xc7894c)
#20 0x00007ff48c49e2ae clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x24fe2ae)
#21 0x00007ff48c4660f1 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x24c60f1)
#22 0x00007ff48c46633e clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x24c633e)
#23 0x00007ff48c48239c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/usr/lib/llvm-18/bin/../lib/libclang-cpp.so.18+0x24e239c)
#24 0x0000561c05dda25c clang_main(int, char**, llvm::ToolContext const&) (/usr/lib/llvm-18/bin/clang+0x1025c)
#25 0x0000561c05de7f32 main (/usr/lib/llvm-18/bin/clang+0x1df32)
#26 0x00007ff48177fc87 __libc_start_main /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:344:0
#27 0x0000561c05dd71ea _start (/usr/lib/llvm-18/bin/clang+0xd1ea)
clang-18: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Ubuntu clang version 18.0.0 (++20231018091808+48a53509e851-1~exp1~20231018091910.1571)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang-18: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-18: note: diagnostic msg: /tmp/mutant-d67eef.c
clang-18: note: diagnostic msg: /tmp/mutant-d67eef.sh
clang-18: note: diagnostic msg: 

********************
```

### PR DESCRIPTION
**Pull request by OCHyams**
Fix https://github.com/llvm/llvm-project/issues/74189 (crash report).

The pruning code uses a BitVector to track which parts of a variable have been
defined in order to find redundant debug records. BitVector uses a u32 to track
size; variable of types with a bit-size greater than max(u32) ish* can't be
represented using a BitVector.

Fix the assertion by introducing a limit on type size. Improve performance by
bringing the limit down to a sensible number and tracking byte-sizes instead
of bit-sizes.

Skipping variables in this pruning code doesn't cause debug info correctness issues;
it just means there may be some extra redundant debug records.

(*) `max(u32) - 63` due to BitVector::NumBitWords implementation.

### CODE REVIEW

**Review by jryans — DISMISSED** [review]
Overall reasonable, but see comments below.

Also, there's a typo in the commit and PR title: "AssignemntTracking" should be "AssignmentTracking"

**Comment by nikic** [comment]
Independently of the crash, doesn't this mean that this code will be incredibly slow for large types?

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 1]
```diff
@@ -2296,8 +2296,11 @@ removeRedundantDbgLocsUsingBackwardScan(const BasicBlock *BB,
           getAggregate(FnVarLocs.getVariable(RIt->VariableID));
       uint64_t SizeInBits = Aggr.first->getSizeInBits().value_or(0);
 
-      if (SizeInBits == 0) {
+      const uint64_t MaxSize = std::numeric_limits<unsigned>::max() - 63;
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp:2299]: Can do - I'll change this if we keep the cutoff tied to bitvector.

**Comment by OCHyams** [comment]
Thanks both for the reviews.

@nikic 

> Independently of the crash, doesn't this mean that this code will be incredibly slow for large types?

There's probably a point at which pruning becomes unprofitable. We could lower the threshold - do you have a ballpark figure you'd suggest?

**Comment by jmorse** [comment]
This sounds good, while agreeing with the comments so far. It's tempting to say that "anyone who's allocating a massively sized stack variable is asking for trouble" and to not support that situation well. Unfortunately that's also exactly the scenario where developers probably want their debugger to work well. What are the implications of over-sized types being ignored by this code, are there any correctness issues or is it just efficiency?

**Comment by OCHyams** [comment]
> What are the implications of over-sized types being ignored by this code, are there any correctness issues or is it just efficiency?

Just efficiency.

As well as choosing a smaller size limit, another easy win here could be to use byte sizes rather than bit sizes, which would bring down the bitvector sizes in all cases (at the cost accidentally not pruning where some fragments don't fit into bytes exactly, but that'll probably be a very rare situation and doesn't impose any correctness problems).

I can also have a look to see if there's something else we can use. Ideally we actually want something like an interval map, though IntervalMap as it exists now is not useful for this since it doesn't support overlapping insertion (and also might be a bit heavy weight too).

**Review by jryans — COMMENTED** [review]
Thanks, the issues I mentioned appear to be resolved. I have no opinion on the details of exactly what cutoff to use, so I'll leave that for others to weigh in on.

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 2]
```diff
@@ -2295,33 +2295,41 @@ removeRedundantDbgLocsUsingBackwardScan(const BasicBlock *BB,
       DebugAggregate Aggr =
           getAggregate(FnVarLocs.getVariable(RIt->VariableID));
       uint64_t SizeInBits = Aggr.first->getSizeInBits().value_or(0);
+      uint64_t SizeInBytes = SizeInBits + 7 / 8;
```
Comment (nikic) [inline llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp:2298]: divideCeil

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 3]
```diff
@@ -2295,33 +2295,41 @@ removeRedundantDbgLocsUsingBackwardScan(const BasicBlock *BB,
...
+          VariableDefinedBytes.try_emplace(Aggr, BitVector(SizeInBytes));
       bool FirstDefinition = InsertResult.second;
-      BitVector &DefinedBits = InsertResult.first->second;
+      BitVector &DefinedBytes = InsertResult.first->second;
 
       DIExpression::FragmentInfo Fragment =
           RIt->Expr->getFragmentInfo().value_or(
               DIExpression::FragmentInfo(SizeInBits, 0));
       bool InvalidFragment = Fragment.endInBits() > SizeInBits;
+      uint64_t StartInBytes = Fragment.startInBits() + 7 / 8;
```
Comment (nikic) [inline llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp:2325]: Not familiar with this code, but shouldn't the start be rounded down?

**Comment by OCHyams** [comment]
+  Address outstanding feedback
+ Update summary

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 4]
```diff
@@ -2295,33 +2295,41 @@ removeRedundantDbgLocsUsingBackwardScan(const BasicBlock *BB,
...
 
-      if (SizeInBits == 0) {
+      // Cutoff for large variables to prevent expensive bitvector operations.
+      const uint64_t MaxSizeBytes = 2048;
+
+      if (SizeInBytes == 0 || SizeInBytes > MaxSizeBytes) {
         // If the size is unknown (0) then keep this location def to be safe.
+        // Do the same for defs of large variables, which would be expensive
+        // to represent with a BitVector.
+        // FIXME: Don't use a BitVector.
```
Comment (SLTozer) [inline llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp:2307]: Is there a reason to not use a BitVector here, given that as discussed we don't lose correctness, and for the vast majority of cases we are well below this limit?

**Review by SLTozer — APPROVED** [review]
Got a small question about the BitVector FIXME, but otherwise this LGTM.

**Pull request merged**